packages/core/src/modules/presentation-exchange/PresentationExchangeService.ts

```diff
@@ -0,0 +1,477 @@
+import {
+  selectCredentialsForRequest,
+  getSphereonOriginalVerifiableCredential,
+  getSphereonW3cVerifiablePresentation,
+  getW3cVerifiablePresentationInstance,
+} from './utils'
+
+export type ProofStructure = Record<string, Record<string, Array<W3cVerifiableCredential>>>
+export type PresentationDefinition = IPresentationDefinition
+
+@injectable()
+export class PresentationExchangeService {
+  private pex = new PEX()
+
+  public async selectCredentialsForRequest(
+    agentContext: AgentContext,
+    presentationDefinition: PresentationDefinition
+  ): Promise<PresentationSubmission> {
+    const credentialRecords = await this.queryCredentialForPresentationDefinition(agentContext, presentationDefinition)
+
+    const didsApi = agentContext.dependencyManager.resolve(DidsApi)
+    const didRecords = await didsApi.getCreatedDids()
+    const holderDids = didRecords.map((didRecord) => didRecord.did)
+
+    return selectCredentialsForRequest(presentationDefinition, credentialRecords, holderDids)
+  }
+
+  /**
+   * Queries the wallet for credentials that match the given presentation definition. This only does an initial query based on the
+   * schema of the input descriptors. It does not do any further filtering based on the constraints in the input descriptors.
+   */
+  private async queryCredentialForPresentationDefinition(
+    agentContext: AgentContext,
+    presentationDefinition: PresentationDefinition
+  ): Promise<Array<W3cCredentialRecord>> {
+    const w3cCredentialRepository = agentContext.dependencyManager.resolve(W3cCredentialRepository)
+    const query: Array<Query<W3cCredentialRecord>> = []
+    const presentationDefinitionVersion = PEX.definitionVersionDiscovery(presentationDefinition)
+
+    if (!presentationDefinitionVersion.version) {
+      throw new PresentationExchangeError(
+        `Unable to determine the Presentation Exchange version from the presentation definition. ${
+          presentationDefinitionVersion.error ?? 'Unknown error'
+        }`
+      )
+    }
+
+    if (presentationDefinitionVersion.version === PEVersion.v1) {
+      const pd = presentationDefinition as PresentationDefinitionV1
+
+      // The schema.uri can contain either an expanded type, or a context uri
+      for (const inputDescriptor of pd.input_descriptors) {
+        for (const schema of inputDescriptor.schema) {
+          query.push({
+            $or: [{ expandedType: [schema.uri] }, { contexts: [schema.uri] }, { type: [schema.uri] }],
+          })
+        }
+      }
+    } else if (presentationDefinitionVersion.version === PEVersion.v2) {
+      // FIXME: As PE version 2 does not have the `schema` anymore, we can't query by schema anymore.
+      // For now we retrieve ALL credentials, as we did the same for V1 with JWT credentials. We probably need
+      // to find some way to do initial filtering, hopefully if there's a filter on the `type` field or something.
```

Comment:
We probably need to look at this in the near future. This will cause issues for sure